Release lite models

guillaumebrg opened the pull request:
This PR is preparing the release of our lite models. It introduces a few concepts that should make it easier for the user to switch between different models when running a demo. Demos that can be run by multiple models now includesa section like this at the top of the script:
```
SUPPORTED_MODEL_CONFIGURATIONS = [
    ModelConfig('StridedInflatedMobileNetV2', 'pro', ['met_converter']),
    ModelConfig('StridedInflatedEfficientNet', 'pro', ['met_converter']),
    ModelConfig('StridedInflatedMobileNetV2', 'lite', ['met_converter']),
    ModelConfig('StridedInflatedEfficientNet', 'lite', ['met_converter']),
]
```
indicating which models are supported by this script. It is possible to select which model to use from the command line (see `--model_name` and `--model_version` arguments). When these arguments are not used (default behavior), it will loop over `SUPPORTED_MODEL_CONFIGURATIONS`, from top to bottom, until it can find a model configuration for which the user has all necessary files. With the example above, this means that the script will try to load "pro" weights first and will only fallback to "lite" weights if "pro" weights were not found.

Please, also check out `models.yaml`. This is my suggestion to map classifier weights to their corresponding backbone weights. This is what we will have to edit when we add support to another demo.

I've left a bunch of TODOs here and there. We should also edit the README I guess.

antoinemrcr commented on scripts/run_gesture_recognition.py:
```diff
@@ -7,46 +7,66 @@
   --title=TITLE              This adds a title to the window display
+  --model_name=NAME          Name of the model to be used.
+  --model_version=VERSION    Version of the model to be used.
+  --use_gpu                  Whether to run inference on the GPU or not.
```
should we default to one model version ?

guillaumebrg replied:
It's not necessary:
> When these arguments are not used (default behavior), it will loop over SUPPORTED_MODEL_CONFIGURATIONS, from top to bottom, until it can find a model configuration for which the user has all necessary files

Adding a default model_name/model_version wouldn't add anything

sunny-panchal commented on sense/loading.py:
```diff
@@ -0,0 +1,125 @@
+from typing import Optional
+from typing import Tuple
+
+with open(os.path.join(os.path.dirname(__file__), 'models.yml')) as f:  # TODO: check if this is the best way of doing this
```
@corneliusboehm any thoughts on this?

corneliusboehm replied:
Apparently you can get more complex (https://stackoverflow.com/a/4060259/6611214), but this looks ok to me

sunny-panchal reviewed (commented):
Finished making the remaining changes and it works as intended :+1:

guillaumebrg commented:
Will fix conflicts tomorrow

guillaumebrg commented on examples/run_fitness_rep_counter.py:
```diff
@@ -7,25 +7,36 @@
   --title=TITLE                   This adds a title to the window display
+  --model_name=NAME          Name of the model to be used.
+  --model_version=VERSION    Version of the model to be used.
+  --use_gpu                  Whether to run inference on the GPU or not.
```
Misalignment

guillaumebrg commented on examples/run_fitness_rep_counter.py:
```diff
@@ -35,22 +46,28 @@
+        model_version
+    )
+
+    # Load backbone extractor
```
"backbone extractor"

guillaumebrg commented on examples/run_fitness_tracker.py:
```diff
@@ -45,29 +60,34 @@
+        model_version
+    )
+
+    # Load backbone extractor
```
"backbone extractor"

guillaumebrg commented on examples/run_gesture_recognition.py:
```diff
@@ -7,50 +7,69 @@
+        model_version
+    )
+
+    # Load backbone extractor
```
backbone extractor

guillaumebrg commented on sense/loading.py:
```diff
@@ -0,0 +1,175 @@
+    """
+    Load backbone weights from a checkpoint file, unless Travis is used. Raises an error pointing
+    to the SDK page in case weights
+    are missing.
```
Misalignment

guillaumebrg commented on sense/loading.py:
```diff
@@ -0,0 +1,175 @@
+import json
```
TODO: add unittests to cover classes/functions from this file

guillaumebrg reviewed (commented):
A bunch of notes to myself... will address these soon

corneliusboehm commented on sense/loading.py:
```diff
@@ -0,0 +1,174 @@
+    :param checkpoint_path:
+        A string representing the absolute/relative path to the checkpoint file.
+    """
+    checkpoint_path = os.path.join(RESOURCES_DIR, checkpoint_path.split(f'resources{os.sep}')[-1])
```
One more reason to remove `resources/` from all paths in the config file

corneliusboehm commented on sense/loading.py:
```diff
@@ -0,0 +1,174 @@
+                                'instructions.'.format(checkpoint_path))
+
+
+def load_backbone_weights(checkpoint_path: str):
+    """
+    Load backbone weights from a checkpoint file, unless Travis is used. Raises an error pointing
+    to the SDK page in case weights are missing.
+
+    :param checkpoint_path:
+        A string representing the absolute/relative path to the checkpoint file.
+    """
+    if not using_travis():
+        return load_weights_from_resources(checkpoint_path)
+    else:
+        print('Weights are not loaded on Travis.')
```
Is it really necessary to have this method wrap `load_weights_from_resources`?

guillaumebrg replied:
It is useful at the time where I write this. Integration tests would complain on Travis without this. But maybe this can go when we refactor `get_relevant_weights`

corneliusboehm commented on sense/loading.py:
```diff
@@ -0,0 +1,174 @@
+    return backbone_network
+
+
+def using_travis():
+    """
+    Returns True if Travis is currently being used.
+    """
+    return os.getenv('TRAVIS', False) == 'true'
```
How about making this a constant `TRAVIS_ENV` on top of the file? Or does it change at runtime?

corneliusboehm reviewed (commented):
Looking good :slightly_smiling_face: Some general feedback that we could address now or later:

I would be in favor of empowering the `ModelConfig` class a little more, e.g. by moving some parts from `get_relevant_weights` or even the model building into it.

Also it would be nice if we could simplify the model setup API even more, because currently this much code is still required to setup a model with a default config and we have it as kind of boilerplate in each example script:
```python
    # Load weights
    selected_config, weights = get_relevant_weights(
        SUPPORTED_MODEL_CONFIGURATIONS,
        model_name,
        model_version
    )

    # Load backbone network
    backbone_network = build_backbone_network(selected_config, weights['backbone'])

    # Create a logistic regression classifier
    gesture_classifier = LogisticRegression(num_in=backbone_network.feature_dim,
                                            num_out=30)
    gesture_classifier.load_state_dict(weights['gesture_recognition'])
    gesture_classifier.eval()

    # Concatenate backbone network and logistic regression
    net = Pipe(backbone_network, gesture_classifier)
```
I'm imagining shortening it to this, which would only require us to store the class and dimensions of the classifier in the config as well:
```python
    # Setup model
    net = load_model(
        SUPPORTED_MODEL_CONFIGURATIONS,
        model_name,
        model_version
    )
```

guillaumebrg commented on sense/models.yml:
```diff
@@ -0,0 +1,29 @@
+    backbone: resources/backbone/strided_inflated_efficientnet_lite.ckpt
+    fitness_activity_recognition: resources/fitness_activity_recognition/strided_inflated_efficientnet_lite_logistic_regression.ckpt
+    gesture_recognition: resources/gesture_recognition/strided_inflated_efficientnet_lite_logistic_regression.ckpt
+    met_converter: resources/calorie_estimation/strided_inflated_efficientnet_lite_met_converter.ckpt
```
Suggested a few changes here: 5ea2f67

guillaumebrg commented on sense/loading.py:
```diff
@@ -0,0 +1,174 @@
+        else:
+            print(f'Could not find at least one of the following files:\n{path_weights_string}')
+
+    raise Exception('ERROR - Weights file missing. To download, please go to '
+                    'https://20bn.com/licensing/sdk/evaluation and follow the '
+                    'instructions.')
```
I think this should be fixed now - `load_weights` was used somewhere, instead of `load_weights_from_resources`

guillaumebrg commented:
@corneliusboehm Thanks for the review! I've addressed the issues that were easy to fix.

Moving `get_relevant_weights`  into `ModelConfig` would probably make sense and I think there are a few things that could be simplified with regards to how we handle loading weights on Travis (not sure we need this many `running_on_travis` conditions)...

I haven't made up my mind yet about your `load_model` suggestion. One disadvantage I see is that it would hide high-level information about the architecture (e.g. "the model for this example is simply a backbone network + a fully connected layer with X classes") and, as a user, I kind of liked that this was exposed in the example scripts. Not sure... Maybe we can discuss it with a larger round of people and see what they prefer?

corneliusboehm commented on sense/loading.py:
```diff
@@ -0,0 +1,183 @@
+    """
+    Prepend the absolute resources path to the provided path.
+    """
+    return os.path.join(RESOURCES_DIR, checkpoint_path.split(f'resources{os.sep}')[-1])
```
The split on `resources/` should not necessary, now that  the paths no longer contain it.

corneliusboehm reviewed (approved):
Thanks for the changes. I confirmed that everything is working :+1: 

We can talk about the other ideas in a larger round.

corneliusboehm commented on tools/conversion/convert_to_tflite.py:
```diff
@@ -61,17 +64,17 @@
         "placeholder_values": {"NUM_CLASSES": "81"},
-        "weights_file": "resources/fitness_activity_recognition/efficientnet_logistic_regression.ckpt",
-        "corresponding_backbone": "efficientnet",
+        "weights_file": MODELS[EFFICIENTNET]['pro'],
```
```suggestion
        "weights_file": MODELS[EFFICIENTNET]['pro']['fitness_activity_recognition'],
```

guillaumebrg replied:
Thanks!
